File: odroid/Share/src/bert/share/xml/XMLUtility.java
/**
 * Copyright 2019. Charles Coughlin. All Rights Reserved.
 *                 MIT License.
 */
package bert.share.xml;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.util.logging.Logger;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;

import org.w3c.dom.Document;
import org.w3c.dom.NamedNodeMap;
import org.w3c.dom.Node;
import org.xml.sax.SAXException;

/**
 * A class with static utility methods dealing with XML files. These methods 
 * are typically designed to return an error string, where a null implies success.
 */
public class XMLUtility {
	private static final String CLSS = "XMLUtility";
	private static final Logger LOGGER = Logger.getLogger(CLSS);
	
	
	public static Document documentFromBytes(byte[] bytes) {
		Document xml = null;
		DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
	    factory.setNamespaceAware(true);
	    try {
	    	DocumentBuilder builder = factory.newDocumentBuilder();
	    	xml = builder.parse(new ByteArrayInputStream(bytes));
	    }
	    catch(ParserConfigurationException pce) {
	    	LOGGER.warning(String.format("%s.documentFromBytes: Failed to create builder (%s)",CLSS,pce.getLocalizedMessage()));
	    }
	    catch(SAXException saxe) {
	    	LOGGER.warning(String.format("%s.documentFromBytes: Illegal XML document (%s)",CLSS,saxe.getLocalizedMessage()));
	    }
	    catch(IOException ioe) {
	    	LOGGER.warning(String.format("%s.documentFromBytes: IOException parsing XML (%s)",CLSS,ioe.getLocalizedMessage()));
	    }
	    
	    return xml;
	}
	
	// =========================  Helper Functions ==============================
	public static String attributeValue(Node element,String name) {
		String value = "";
		NamedNodeMap attributes = element.getAttributes();
		Node node = attributes.getNamedItem(name);
		if( node!=null ) value = node.getNodeValue();
		return value;
	}

}

File: Bert/common/src/main/kotlin/chuckcoughlin/bert/common/util/XMLUtility.kt
/**
 * Copyright 2022. Charles Coughlin. All Rights Reserved.
 * MIT License.
 */
package chuckcoughlin.bert.common.util

import org.w3c.dom.Document
import org.w3c.dom.Node
import org.xml.sax.SAXException
import java.io.ByteArrayInputStream
import java.io.IOException
import java.util.logging.Logger
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.parsers.ParserConfigurationException

/**
 * A class with static utility methods dealing with XML files. These methods
 * are typically designed to return an error string, where a null implies success.
 */
object XMLUtility {
    private const val CLSS = "XMLUtility"
    private val LOGGER = Logger.getLogger(CLSS)
    fun documentFromBytes(bytes: ByteArray): Document {
        val factory = DocumentBuilderFactory.newInstance()
        factory.isNamespaceAware = true
        val builder = factory.newDocumentBuilder()
        var xml = builder.newDocument()
        try {
            xml = builder.parse(ByteArrayInputStream(bytes))
        }
        catch (pce: ParserConfigurationException) {
            LOGGER.warning(String.format("%s.documentFromBytes: Failed to create builder (%s)",
                    CLSS,pce.localizedMessage))
        }
        catch (saxe: SAXException) {
            LOGGER.warning(String.format("%s.documentFromBytes: Illegal XML document (%s)",
                    CLSS,saxe.localizedMessage))
        }
        catch (ioe: IOException) {
            LOGGER.warning(String.format("%s.documentFromBytes: IOException parsing XML (%s)",
                    CLSS,ioe.localizedMessage))
        }
        return xml
    }

    // =========================  Helper Functions ==============================
    fun attributeValue(element: Node, name: String): String {
        var value:String  = ""
        val attributes = element.attributes
        val node = attributes.getNamedItem(name)
        if (node != null) {
            if( node.nodeValue!=null) {
                value = node.nodeValue
            }
        }
        return value
    }
}